src/analysis/cache.Rmd

```diff
@@ -0,0 +1,313 @@
+
+In the table below: real_time is wallclock time, sd stands for the standard deviation and rel stands for relative deviation, perf refers to exectuion with perf and pure refers to execution witout perf.
+
+```{r}
+t
+```
+```{r}
+t <- aggregate(cbind(time_measurements$user_time_perf,time_measurements$sys_time_perf,time_measurements$user_time_pure,time_measurements$sys_time_pure), by = list(time_measurements$opcode, time_measurements$op_count), FUN = mean)
+
+colnames(t) <- c("opcode", "op_count", "user_time_perf", "sys_time_perf", "user_time_pure", "sys_time_pure")
+
+mt <- sqldf("SELECT opcode, max(abs(user_time_perf-user_time_pure)/user_time_pure) as max_increase FROM t GROUP BY opcode")
+```
+
+We calculate maximal relative increase of CPU time usage when using perf. For most opcodes it is around 0.1%. There is one opcode with 0.5% increase. These are very low values and time executions are almost the same when using perf and not using perf.
+
+```{r}
+mt
+```
+
+The detailed presentations.
+
+```{r fig.width=10}
+pp <- function(tt) {
+  opcode <- tt[1,1]
+  m <- max(tt$user_time_perf)
+  plot(tt$op_count, tt$user_time_perf, col=1, type="p", ylim=c(0,m+1), xlab="op_count", ylab="sys_time/user_time", main=opcode)
+  lines(tt$op_count, tt$user_time_pure, col=2, type="p")
+  lines(tt$op_count, tt$sys_time_perf, col=3, type="p")
+  lines(tt$op_count, tt$sys_time_pure, col=4, type="p")
+}
+for (it in split(t, f=t$opcode)) {
+  pp(it)
+}
+```
+
+It is hard to distinguish the perf times and pure times as the graph points overlap. Concluding, the impact of using perf is negligible for our analysis.
+
+## Analysis
+
+We use trimmed means (10% cut off) of measured stats for analysis. The stats are measured with perf tool:
+
+- task_clock
+- context_switches
+- page_faults
+- instructions
+- branches
+- branch_misses
+- L1_dcache_loads
+- LLC_loads
+- LLC_load_misses
+- L1_icache_loads
+- L1_icache_load_misses
+- dTLB_loads
+- dTLB_load_misses
+- iTLB_loads
+- iTLB_load_misses
+
+```{r}
+programs = read.csv("../../local/pg_marginal_50.csv")
+measurements_data = read.csv("../../local/perf_evmone_pg_marginal_50_1000000.csv")
+measurements = sqldf("SELECT opcode, op_count, 
+                  measurements_data.*
+          FROM measurements_data
+          INNER JOIN programs ON(measurements_data.program_id = programs.program_id)")
+measurements_avg <- aggregate(cbind(measurements$task_clock,measurements$context_switches,measurements$page_faults,measurements$instructions,measurements$branches,measurements$branch_misses,measurements$L1_dcache_loads,measurements$LLC_loads,measurements$LLC_load_misses,measurements$L1_icache_loads,measurements$L1_icache_load_misses,measurements$dTLB_loads,measurements$dTLB_load_misses,measurements$iTLB_loads,measurements$iTLB_load_misses), by = list(measurements$opcode, measurements$op_count), FUN = mean, trim=0.1)
```

Comment:
I'm not 100% sure, but maybe you can rewrite the `cbind` to something like: `measurements[, c('task_clock', 'context_switches', ...)]` should be much more succinct. Possibly, the `colnames() <-...` will not be necessary anymore.